File: mycollab-services-community/src/test/java/com/esofthead/mycollab/module/project/service/ProjectFollowingTicketServiceTest.java
/**
 * This file is part of mycollab-services-community.
 *
 * mycollab-services-community is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-services-community is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-services-community.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.project.service;

import com.esofthead.mycollab.core.arguments.NumberSearchField;
import com.esofthead.mycollab.core.arguments.SearchRequest;
import com.esofthead.mycollab.core.arguments.SetSearchField;
import com.esofthead.mycollab.core.arguments.StringSearchField;
import com.esofthead.mycollab.module.project.domain.FollowingTicket;
import com.esofthead.mycollab.module.project.domain.criteria.FollowingTicketSearchCriteria;
import com.esofthead.mycollab.test.DataSet;
import com.esofthead.mycollab.test.service.IntergrationServiceTest;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.tuple;

@RunWith(SpringJUnit4ClassRunner.class)
public class ProjectFollowingTicketServiceTest extends IntergrationServiceTest {

    private static final DateFormat DATE_FORMAT = new SimpleDateFormat(
            "yyyy-MM-dd hh:mm:ss");

    @Autowired
    private ProjectFollowingTicketService projectFollowingTicketService;

    private FollowingTicketSearchCriteria getCriteria() {
        FollowingTicketSearchCriteria criteria = new FollowingTicketSearchCriteria();
        criteria.setExtraTypeIds(new SetSearchField<>(1, 2));
        criteria.setSaccountid(new NumberSearchField(1));
        criteria.setUser(StringSearchField.and("hainguyen@esofthead.com"));
        return criteria;
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicket() throws ParseException {
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(
                        getCriteria(), 0, Integer.MAX_VALUE));
        assertThat(projectFollowingTickets).extracting("type", "summary",
                "monitorDate").contains(
                tuple("Project-Task", "task 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Task", "task 2", DATE_FORMAT.parse("2014-10-22 00:00:00")),
                tuple("Project-Bug", "bug 1", DATE_FORMAT.parse("2014-10-23 00:00:00")),
                tuple("Project-Bug", "bug 2", DATE_FORMAT.parse("2014-10-24 00:00:00")),
                tuple("Project-Task", "task 3", DATE_FORMAT.parse("2014-09-21 00:00:00")),
                tuple("Project-Task", "task 4", DATE_FORMAT.parse("2014-09-22 00:00:00")),
                tuple("Project-Bug", "bug 3", DATE_FORMAT.parse("2014-09-23 00:00:00")),
                tuple("Project-Bug", "bug 4", DATE_FORMAT.parse("2014-09-24 00:00:00")),
                tuple("Project-Problem", "problem 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Problem", "problem 2", DATE_FORMAT.parse("2014-10-22 00:00:00")),
                tuple("Project-Risk", "risk 1", DATE_FORMAT.parse("2014-10-23 00:00:00")),
                tuple("Project-Risk", "risk 2", DATE_FORMAT.parse("2014-10-24 00:00:00")),
                tuple("Project-Problem", "problem 3", DATE_FORMAT.parse("2014-09-21 00:00:00")),
                tuple("Project-Problem", "problem 4", DATE_FORMAT.parse("2014-09-22 00:00:00")),
                tuple("Project-Risk", "risk 3", DATE_FORMAT.parse("2014-09-23 00:00:00")),
                tuple("Project-Risk", "risk 4", DATE_FORMAT.parse("2014-09-24 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(16);
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicketBySummary()
            throws ParseException {
        FollowingTicketSearchCriteria criteria = getCriteria();
        criteria.setSummary(StringSearchField.and("1"));
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(criteria, 0, Integer.MAX_VALUE));
        assertThat(projectFollowingTickets).extracting("type", "summary", "monitorDate").contains(
                tuple("Project-Task", "task 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Bug", "bug 1", DATE_FORMAT.parse("2014-10-23 00:00:00")),
                tuple("Project-Problem", "problem 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Risk", "risk 1", DATE_FORMAT.parse("2014-10-23 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(4);
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicketOfTaskAndBug()
            throws ParseException {
        FollowingTicketSearchCriteria criteria = getCriteria();
        criteria.setTypes(new SetSearchField<>("Project-Task", "Project-Bug"));
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(criteria, 0, Integer.MAX_VALUE));
        assertThat(projectFollowingTickets).extracting("type", "summary",
                "monitorDate").contains(
                tuple("Project-Task", "task 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Task", "task 2", DATE_FORMAT.parse("2014-10-22 00:00:00")),
                tuple("Project-Task", "task 3", DATE_FORMAT.parse("2014-09-21 00:00:00")),
                tuple("Project-Task", "task 4", DATE_FORMAT.parse("2014-09-22 00:00:00")),
                tuple("Project-Bug", "bug 1", DATE_FORMAT.parse("2014-10-23 00:00:00")),
                tuple("Project-Bug", "bug 2", DATE_FORMAT.parse("2014-10-24 00:00:00")),
                tuple("Project-Bug", "bug 3", DATE_FORMAT.parse("2014-09-23 00:00:00")),
                tuple("Project-Bug", "bug 4", DATE_FORMAT.parse("2014-09-24 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(8);
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicketOfTask() throws ParseException {
        FollowingTicketSearchCriteria criteria = getCriteria();
        criteria.setType(StringSearchField.and("Project-Task"));
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(criteria, 0, Integer.MAX_VALUE));
        assertThat(projectFollowingTickets).extracting("type", "summary",
                "monitorDate").contains(
                tuple("Project-Task", "task 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Task", "task 2", DATE_FORMAT.parse("2014-10-22 00:00:00")),
                tuple("Project-Task", "task 3", DATE_FORMAT.parse("2014-09-21 00:00:00")),
                tuple("Project-Task", "task 4", DATE_FORMAT.parse("2014-09-22 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(4);
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicketOfRisk() throws ParseException {
        FollowingTicketSearchCriteria criteria = getCriteria();
        criteria.setType(StringSearchField.and("Project-Risk"));
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(criteria, 0, Integer.MAX_VALUE));

        assertThat(projectFollowingTickets).extracting("type", "summary",
                "monitorDate").contains(
                tuple("Project-Risk", "risk 1", DATE_FORMAT.parse("2014-10-23 00:00:00")),
                tuple("Project-Risk", "risk 2", DATE_FORMAT.parse("2014-10-24 00:00:00")),
                tuple("Project-Risk", "risk 3", DATE_FORMAT.parse("2014-09-23 00:00:00")),
                tuple("Project-Risk", "risk 4", DATE_FORMAT.parse("2014-09-24 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(4);
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicketOfProblem()
            throws ParseException {
        FollowingTicketSearchCriteria criteria = getCriteria();
        criteria.setType(StringSearchField.and("Project-Problem"));
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(criteria, 0, Integer.MAX_VALUE));

        assertThat(projectFollowingTickets).extracting("type", "summary",
                "monitorDate").contains(
                tuple("Project-Problem", "problem 1", DATE_FORMAT.parse("2014-10-21 00:00:00")),
                tuple("Project-Problem", "problem 2", DATE_FORMAT.parse("2014-10-22 00:00:00")),
                tuple("Project-Problem", "problem 3", DATE_FORMAT.parse("2014-09-21 00:00:00")),
                tuple("Project-Problem", "problem 4", DATE_FORMAT.parse("2014-09-22 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(4);
    }

    @SuppressWarnings("unchecked")
    @DataSet
    @Test
    public void testGetListProjectFollowingTicketOfBug() throws ParseException {
        FollowingTicketSearchCriteria criteria = getCriteria();
        criteria.setType(StringSearchField.and("Project-Bug"));
        List<FollowingTicket> projectFollowingTickets = projectFollowingTicketService
                .findPagableListByCriteria(new SearchRequest<>(criteria, 0, Integer.MAX_VALUE));

        assertThat(projectFollowingTickets).extracting("type", "summary",
                "monitorDate").contains(
                tuple("Project-Bug", "bug 1", DATE_FORMAT.parse("2014-10-23 00:00:00")),
                tuple("Project-Bug", "bug 2", DATE_FORMAT.parse("2014-10-24 00:00:00")),
                tuple("Project-Bug", "bug 3", DATE_FORMAT.parse("2014-09-23 00:00:00")),
                tuple("Project-Bug", "bug 4", DATE_FORMAT.parse("2014-09-24 00:00:00")));
        assertThat(projectFollowingTickets.size()).isEqualTo(4);
    }

    @DataSet
    @Test
    public void testGetTotalCount() throws ParseException {
        assertThat(projectFollowingTicketService.getTotalCount(getCriteria())).isEqualTo(16);
    }
}

File: mycollab-services-community/src/test/java/com/mycollab/module/project/service/ProjectFollowingTicketServiceTest.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.project.service

import com.mycollab.db.arguments.BasicSearchRequest
import com.mycollab.db.arguments.NumberSearchField
import com.mycollab.db.arguments.SetSearchField
import com.mycollab.db.arguments.StringSearchField
import com.mycollab.module.project.domain.FollowingTicket
import com.mycollab.module.project.domain.criteria.FollowingTicketSearchCriteria
import com.mycollab.test.DataSet
import com.mycollab.test.rule.DbUnitInitializerRule
import com.mycollab.test.spring.IntegrationServiceTest
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.tuple
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.text.ParseException

@ExtendWith(SpringExtension::class, DbUnitInitializerRule::class)
class ProjectFollowingTicketServiceTest : IntegrationServiceTest() {

    @Autowired
    private lateinit var projectFollowingTicketService: ProjectFollowingTicketService

    private val criteria: FollowingTicketSearchCriteria
        get() {
            val criteria = FollowingTicketSearchCriteria()
            criteria.extraTypeIds = SetSearchField(1, 2)
            criteria.saccountid = NumberSearchField(1)
            criteria.user = StringSearchField.and("hainguyen@esofthead.com")
            return criteria
        }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetListProjectFollowingTicket() {
        val projectFollowingTickets = projectFollowingTicketService.findPageableListByCriteria(BasicSearchRequest(criteria)) as List<FollowingTicket>
        assertThat<FollowingTicket>(projectFollowingTickets).extracting("type", "name").contains(
                tuple("Project-Task", "task 1"),
                tuple("Project-Task", "task 2"),
                tuple("Project-Bug", "bug 1"),
                tuple("Project-Bug", "bug 2"),
                tuple("Project-Task", "task 3"),
                tuple("Project-Task", "task 4"),
                tuple("Project-Bug", "bug 3"),
                tuple("Project-Bug", "bug 4"),
                tuple("Project-Risk", "risk 1"),
                tuple("Project-Risk", "risk 2"),
                tuple("Project-Risk", "risk 3"),
                tuple("Project-Risk", "risk 4"))
        assertThat(projectFollowingTickets.size).isEqualTo(12)
    }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetListProjectFollowingTicketBySummary() {
        val criteria = criteria
        criteria.name = StringSearchField.and("1")
        val projectFollowingTickets = projectFollowingTicketService.findPageableListByCriteria(BasicSearchRequest(criteria)) as List<FollowingTicket>
        assertThat<FollowingTicket>(projectFollowingTickets).extracting("type", "name").contains(
                tuple("Project-Task", "task 1"),
                tuple("Project-Bug", "bug 1"),
                tuple("Project-Risk", "risk 1"))
        assertThat(projectFollowingTickets.size).isEqualTo(3)
    }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetListProjectFollowingTicketOfTaskAndBug() {
        val criteria = criteria
        criteria.types = SetSearchField("Project-Task", "Project-Bug")
        val projectFollowingTickets = projectFollowingTicketService.findPageableListByCriteria(BasicSearchRequest(criteria)) as List<FollowingTicket>
        assertThat<FollowingTicket>(projectFollowingTickets).extracting("type", "name").contains(
                tuple("Project-Task", "task 1"),
                tuple("Project-Task", "task 2"),
                tuple("Project-Task", "task 3"),
                tuple("Project-Task", "task 4"),
                tuple("Project-Bug", "bug 1"),
                tuple("Project-Bug", "bug 2"),
                tuple("Project-Bug", "bug 3"),
                tuple("Project-Bug", "bug 4"))
        assertThat(projectFollowingTickets.size).isEqualTo(8)
    }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetListProjectFollowingTicketOfTask() {
        val criteria = criteria
        criteria.type = StringSearchField.and("Project-Task")
        val projectFollowingTickets = projectFollowingTicketService.findPageableListByCriteria(BasicSearchRequest(criteria)) as List<FollowingTicket>
        assertThat<FollowingTicket>(projectFollowingTickets).extracting("type", "name").contains(
                tuple("Project-Task", "task 1"),
                tuple("Project-Task", "task 2"),
                tuple("Project-Task", "task 3"),
                tuple("Project-Task", "task 4"))
        assertThat(projectFollowingTickets.size).isEqualTo(4)
    }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetListProjectFollowingTicketOfRisk() {
        val criteria = criteria
        criteria.type = StringSearchField.and("Project-Risk")
        val projectFollowingTickets = projectFollowingTicketService.findPageableListByCriteria(BasicSearchRequest(criteria)) as List<FollowingTicket>

        assertThat<FollowingTicket>(projectFollowingTickets).extracting("type", "name").contains(
                tuple("Project-Risk", "risk 1"),
                tuple("Project-Risk", "risk 2"),
                tuple("Project-Risk", "risk 3"),
                tuple("Project-Risk", "risk 4"))
        assertThat(projectFollowingTickets.size).isEqualTo(4)
    }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetListProjectFollowingTicketOfBug() {
        val criteria = criteria
        criteria.type = StringSearchField.and("Project-Bug")
        val projectFollowingTickets = projectFollowingTicketService.findPageableListByCriteria(BasicSearchRequest(criteria)) as List<FollowingTicket>

        assertThat<FollowingTicket>(projectFollowingTickets).extracting("type", "name").contains(
                tuple("Project-Bug", "bug 1"),
                tuple("Project-Bug", "bug 2"),
                tuple("Project-Bug", "bug 3"),
                tuple("Project-Bug", "bug 4"))
        assertThat(projectFollowingTickets.size).isEqualTo(4)
    }

    @DataSet
    @Test
    @Throws(ParseException::class)
    fun testGetTotalCount() {
        assertThat(projectFollowingTicketService.getTotalCount(criteria)).isEqualTo(12)
    }
}
